plat/kvm/x86/lxboot.c

```diff
@@ -0,0 +1,161 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2022, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/essentials.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/types.h>
+#include <uk/arch/paging.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/plat/common/lcpu.h>
+#include <uk/plat/common/memory.h>
+#include <uk/plat/common/sections.h>
+#include <kvm-x86/lxboot.h>
+
+#define lxboot_crash(rc, msg, ...) ukplat_crash()
+
+void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi);
+
+static void
+lxboot_init_cmdline(struct ukplat_bootinfo *bi, struct lxboot_params *bp)
+{
+	struct ukplat_memregion_desc mrd = {0};
+	__u64 cmdline_addr;
+	__sz cmdline_size;
+	int rc;
+
+	cmdline_addr = bp->ext_cmd_line_ptr;
+	cmdline_addr <<= 32;
+	cmdline_addr |= bp->hdr.cmd_line_ptr;
+
+	cmdline_size = bp->hdr.cmd_line_size;
+
+	if (cmdline_addr == 0)
+		lxboot_crash(-EINVAL, "Command line address is zero");
+
+	if (cmdline_size == 0)
+		return;
+
+	mrd.pbase = cmdline_addr;
+	mrd.vbase = cmdline_addr;
+	mrd.len = cmdline_size;
+	mrd.type = UKPLAT_MEMRT_CMDLINE;
+	mrd.flags = UKPLAT_MEMRF_READ | UKPLAT_MEMRF_MAP;
+#ifdef CONFIG_UKPLAT_MEMRNAME
+	memcpy(mrd.name, "cmdline", sizeof("cmdline"));
+#endif
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, &mrd);
+	if (rc < 0)
+		lxboot_crash(rc, "Unable to add ram mapping");
+
+	bi->cmdline = cmdline_addr;
+	bi->cmdline_len = cmdline_size;
+}
+
+static void
+lxboot_init_initrd(struct ukplat_bootinfo *bi, struct lxboot_params *bp)
+{
+	__u64 initrd_addr;
+	__sz initrd_size;
+	struct ukplat_memregion_desc mrd = {0};
+	int rc;
+
+	initrd_addr = bp->ext_ramdisk_image;
+	initrd_addr <<= 32;
+	initrd_addr |= bp->hdr.ramdisk_image;
+
+	initrd_size = bp->ext_ramdisk_size;
+	initrd_size <<= 32;
+	initrd_size |= bp->hdr.ramdisk_size;
+
+	if (initrd_addr == 0 || initrd_size == 0)
+		return;
+
+	mrd.type = UKPLAT_MEMRT_INITRD;
+	mrd.flags = UKPLAT_MEMRF_MAP | UKPLAT_MEMRF_READ;
+	mrd.vbase = initrd_addr;
+	mrd.pbase = initrd_addr;
+	mrd.len = initrd_size;
+#ifdef CONFIG_UKPLAT_MEMRNAME
+	memcpy(mrd.name, "initrd", sizeof("initrd"));
+#endif
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, &mrd);
+	if (rc < 0)
+		lxboot_crash(rc, "Unable to add ram mapping");
+}
+
+static void
+lxboot_init_mem(struct ukplat_bootinfo *bi, struct lxboot_params *bp)
+{
+	struct ukplat_memregion_desc mrd = {0};
+	struct lxboot_e820_entry *entry;
+	__u64 start, end;
+	int rc;
+	__s8 i;
+
+	for (i = 0; i < bp->e820_entries; i++) {
```

Comment:
I have just stumbled upon this while working on something else but, IIRC, witth paging disabled, the allocator expects that the free memory regions that it owns are all below 4GiB since that is what the static boot page tables map. Just making sure, but are all of these entries below 4GiB or should this loop here stop when it reaches the first memory region above 4GiB if paging is disabled (assuming they are sorted)?

I think that the multiboot code does not have to worry about that. Although the address field of the entry is 64 bits, the `mem_upper` field is always 32 bits so I guess that allows it to safely assue that no memory region is above 4GiB.

Reply:
That's a good point, but I don't think the boot entry code is the proper place to fix this problem. 
Also, if the point about multiboot is true, wouldn't that also mean that it currently doesn't support more than 4GiB of memory?

Reply:
Multiboot1 does not but the e820 entry does. That's why I said its address field is 64-bit.

Reply:
Oh and I agree the boot entry point is not the place to fix this, you are right. The boot entry point should everything it can to fill the bootinfo struct the best it can. Therefore, perhaps it should be the job of what comes after to accomodate what it is given for what it can offer. That being said, should this check be done in heap_init un ukboot instead?

Reply:
I just talked to Marc and he pointed me to [this](https://github.com/unikraft/unikraft/blob/195cc32f7e5e65d8929ae65b8fa9203ddf246ba6/plat/kvm/x86/setup.c#L233). So the boot code already prunes the high memory areas.

Reply:
Ah I see. I must have missed it.
 I was just making sure. Thanks for checking :).